File: 600/696 - Count Binary Substrings.c
int countBinarySubstrings(char* s) {
    int ret = 0, curr = 0, prev = 0;
    char c, pc = *s;
    
    while (c = *s++) {
        if (c == pc) {
            curr++;
        } else {
            ret += curr > prev ? prev : curr;
            prev = curr;
            curr = 1;
        }
        pc = c;
    }
    
    return ret + (curr > prev ? prev : curr);
}

File: csharp/696.count-binary-substrings.cs
/*
 * @lc app=leetcode id=696 lang=csharp
 *
 * [696] Count Binary Substrings
 */

// @lc code=start
public class Solution {
    public int CountBinarySubstrings(string s) {
        int ret = 0, prev = 0, i = 0, n = s.Length;
        while (i < n) {
            char c = s[i];
            int j;
            for (j = i; j < n && s[j] == c; ++j);
            int count = j - i;
            ret += Math.Min(count, prev);
            prev = count;
            i = j;
        }
        return ret;
    }
}
// @lc code=end

